### PR TITLE
feat(checkbox): add indeterminate support

### PR DESCRIPTION
**Pull request by assuncaocharles**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ yarn change`

#### Description of changes

This PR adds `indeterminate` support for `Checkbox` component 

1FvnEgJYcb

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts [hunk 1]
```diff
@@ -115,9 +117,15 @@ export const checkboxStyles: ComponentSlotStylesPrepared<CheckboxStylesProps, Ch
...
+    ...(p.checked &&
+      !p.indeterminate && {
+        borderColor: v.checkedBorderColor,
+        backgroundImage: checkboxIndicatorUrl(v.checkedIndicatorColor, v.checkedBackground),
+      }),
+
+    ...(p.indeterminate && {
       borderColor: v.checkedBorderColor,
-      backgroundImage: checkboxIndicatorUrl(v.checkedIndicatorColor, v.checkedBackground),
+      backgroundImage: checkboxIndicatorIndeterminatedUrl(v.checkedIndicatorColor, v.checkedBackground),
```
Comment (pompomon) [inline packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts:128]:
```suggestion
      backgroundImage: checkboxIndicatorIndeterminateUrl(v.checkedIndicatorColor, v.checkedBackground),
```
__nit__

Comment (assuncaocharles) [reply packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts:128]: changed

**Review by pompomon — APPROVED** [review]

**Comment by layershifter** [comment]
@assuncaocharles Please update implementation according to spec changes 🙏

**Review by layershifter — CHANGES_REQUESTED** [review]
^

**Comment by assuncaocharles** [comment]
@layershifter can you have a look now?

File: packages/fluentui/CHANGELOG.md [hunk 2]
```diff
@@ -48,6 +48,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Added `disabledFocusable` prop for `Button` component. @jurokapsiar ([#16419](https://github.com/microsoft/fluentui/pull/16419))
 - Added `ChannelShareIcon` @notandrew ([#16619](https://github.com/microsoft/fluentui/pull/16619))
 - Added `FormTextArea` component @assuncaocharles ([#16660](https://github.com/microsoft/fluentui/pull/16660))
+- Added `indeterminate` support for `Checkbox`. @assuncaocharles ([#16081](https://github.com/microsoft/fluentui/pull/16081))
```
Comment (layershifter) [inline packages/fluentui/CHANGELOG.md:51]:
```suggestion
- Added `checked="mixed"` support for `Checkbox` @assuncaocharles ([#16081](https://github.com/microsoft/fluentui/pull/16081))
```

File: packages/fluentui/accessibility/src/behaviors/Checkbox/checkboxBehavior.ts [hunk 3]
```diff
@@ -13,8 +13,11 @@ import { IS_FOCUSABLE_ATTRIBUTE } from '../../attributes';
 export const checkboxBehavior: Accessibility<CheckboxBehaviorProps> = props => ({
   attributes: {
     root: {
-      'aria-checked': !!props.checked,
+      'aria-checked': !!props.indeterminate ? 'mixed' : !!props.checked,
```
Comment (layershifter) [inline packages/fluentui/accessibility/src/behaviors/Checkbox/checkboxBehavior.ts:16]: This is not valid

File: packages/fluentui/accessibility/src/behaviors/Checkbox/checkboxBehavior.ts [hunk 4]
```diff
@@ -32,6 +35,10 @@ export const checkboxBehavior: Accessibility<CheckboxBehaviorProps> = props => (
 export type CheckboxBehaviorProps = {
   /** Whether or not item is checked. */
   checked?: boolean;
+  /** Whether or not item is mixed state. */
+  indeterminate?: boolean;
```
Comment (layershifter) [inline packages/fluentui/accessibility/src/behaviors/Checkbox/checkboxBehavior.ts:39]:
```suggestion
```

File: packages/fluentui/accessibility/src/behaviors/Checkbox/checkboxBehavior.ts [hunk 5]
```diff
@@ -13,8 +13,11 @@ import { IS_FOCUSABLE_ATTRIBUTE } from '../../attributes';
 export const checkboxBehavior: Accessibility<CheckboxBehaviorProps> = props => ({
   attributes: {
     root: {
-      'aria-checked': !!props.checked,
+      'aria-checked': !!props.indeterminate ? 'mixed' : !!props.checked,
       'aria-disabled': props.disabled,
+      ...(!!props.controlsIds && {
```
Comment (layershifter) [inline packages/fluentui/accessibility/src/behaviors/Checkbox/checkboxBehavior.ts:18]: Do we really need this? Can we just use `aria-controls`?

File: packages/fluentui/docs/src/examples/components/Checkbox/BestPractices/CheckboxBestPractices.tsx [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+const doList = ['Add `controlsIds="id1 id2"` for indeterminate checkbox referring to the controls which it depends'];
+
+const dontList = [];
+
+const CardBestPractices = () => {
+  return <ComponentBestPractices doList={doList} dontList={dontList} />;
+};
+
+export default CardBestPractices;
```
Comment (layershifter) [inline packages/fluentui/docs/src/examples/components/Checkbox/BestPractices/CheckboxBestPractices.tsx:13]:
```suggestion
export default CheckboxBestPractices;
```

File: packages/fluentui/docs/src/examples/components/Checkbox/BestPractices/CheckboxBestPractices.tsx [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import * as React from 'react';
+
+import ComponentBestPractices from '../../../../components/ComponentBestPractices';
+
+const doList = ['Add `controlsIds="id1 id2"` for indeterminate checkbox referring to the controls which it depends'];
```
Comment (layershifter) [inline packages/fluentui/docs/src/examples/components/Checkbox/BestPractices/CheckboxBestPractices.tsx:5]: https://github.com/microsoft/fluentui/pull/16081/files#r571040798

File: packages/fluentui/react-northstar/src/components/Checkbox/Checkbox.tsx [hunk 8]
```diff
@@ -53,20 +56,22 @@ export interface CheckboxProps extends UIComponentProps, ChildrenComponentProps
...
+  onClick?: ComponentEventHandler<Omit<CheckboxProps, 'checked'> & { checked: boolean }>;
 
   /** A checkbox can be formatted to show an "on or off" choice. */
   toggle?: boolean;
 }
 
-export type CheckboxStylesProps = Pick<CheckboxProps, 'checked' | 'disabled' | 'labelPosition' | 'toggle'>;
+export type CheckboxStylesProps = Pick<CheckboxProps, 'checked' | 'disabled' | 'labelPosition' | 'toggle'> & {
+  indeterminate: boolean;
+};
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Checkbox/Checkbox.tsx:74]:
```suggestion
export type CheckboxStylesProps = Pick<CheckboxProps, 'checked' | 'disabled' | 'labelPosition' | 'toggle'>;
```

File: packages/fluentui/react-northstar/src/components/Checkbox/Checkbox.tsx [hunk 9]
```diff
@@ -96,17 +101,22 @@ export const Checkbox: ComponentWithAs<'div', CheckboxProps> & FluentComponentSt
...
-    mapPropsToState: () => ({ checked }),
+    mapPropsToState: () => ({ checked: checked === 'mixed' ? false : checked }),
   });
+
   const getA11Props = useAccessibility(props.accessibility, {
     debugName: Checkbox.displayName,
     mapPropsToBehavior: () => ({
       checked: state.checked,
       disabled,
+      indeterminate: checked === 'mixed',
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Checkbox/Checkbox.tsx:118]:
```suggestion
```

File: packages/fluentui/react-northstar/src/components/Checkbox/Checkbox.tsx [hunk 10]
```diff
@@ -116,10 +126,12 @@ export const Checkbox: ComponentWithAs<'div', CheckboxProps> & FluentComponentSt
     },
     rtl: context.rtl,
   });
+
   const { classes, styles: resolvedStyles } = useStyles<CheckboxStylesProps>(Checkbox.displayName, {
     className: checkboxClassName,
     mapPropsToStyles: () => ({
       checked: state.checked,
+      indeterminate: checked === 'mixed',
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/components/Checkbox/Checkbox.tsx:134]:
```suggestion
```

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts [hunk 11]
```diff
@@ -115,9 +117,15 @@ export const checkboxStyles: ComponentSlotStylesPrepared<CheckboxStylesProps, Ch
     backgroundPosition: 'center',
     backgroundRepeat: 'no-repeat',
 
-    ...(p.checked && {
+    ...(p.checked &&
+      !p.indeterminate && {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts:121]:
```suggestion
      {
```

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts [hunk 12]
```diff
@@ -115,9 +117,15 @@ export const checkboxStyles: ComponentSlotStylesPrepared<CheckboxStylesProps, Ch
...
+        borderColor: v.checkedBorderColor,
+        backgroundImage: checkboxIndicatorUrl(v.checkedIndicatorColor, v.checkedBackground),
+      }),
+
+    ...(p.indeterminate && {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts:126]:
```suggestion
    ...(p.checked === 'mixed' && {
```

File: packages/fluentui/docs/src/examples/components/Checkbox/States/CheckboxExampleIndeterminate.shorthand.tsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
+import { Checkbox, CheckboxProps } from '@fluentui/react-northstar';
+import * as React from 'react';
+
+const CheckboxExampleChecked = () => {
+  const [lettuceChecked, setLettuceCheck] = React.useState(false);
+  const [tomatoChecked, setTomatoCheck] = React.useState(false);
```
Comment (layershifter) [inline packages/fluentui/docs/src/examples/components/Checkbox/States/CheckboxExampleIndeterminate.shorthand.tsx:6]:
```suggestion
  const [lettuceChecked, setLettuceChecked] = React.useState(false);
  const [tomatoChecked, setTomatoChecked] = React.useState(false);
```

File: packages/fluentui/docs/src/examples/components/Checkbox/States/CheckboxExampleIndeterminate.shorthand.tsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
+import { Checkbox, CheckboxProps } from '@fluentui/react-northstar';
+import * as React from 'react';
+
+const CheckboxExampleChecked = () => {
```
Comment (layershifter) [inline packages/fluentui/docs/src/examples/components/Checkbox/States/CheckboxExampleIndeterminate.shorthand.tsx:4]:
```suggestion
const CheckboxExampleCheckedMixed = () => {
```

File: packages/fluentui/docs/src/examples/components/Checkbox/States/CheckboxExampleIndeterminate.shorthand.tsx [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  const [lettuceChecked, setLettuceCheck] = React.useState(false);
+  const [tomatoChecked, setTomatoCheck] = React.useState(false);
+  let checked: CheckboxProps['checked'] = lettuceChecked && tomatoChecked;
+  if (!checked) {
+    checked = !lettuceChecked && !tomatoChecked ? false : 'mixed';
+  }
```
Comment (layershifter) [inline packages/fluentui/docs/src/examples/components/Checkbox/States/CheckboxExampleIndeterminate.shorthand.tsx:10]: I played with this example, I suggest to add a third checkbox to a group and use `useReducer()` instead:

```
const [state, dispatch] = ...
/* state: { lettuce, tomato, ?? }
```

To quickly compute state of the main checkbox you can use magic ternary 🌜 

```
const checked = Object.keys(state).every(Boolean) || (Object.keys(state).some(Boolean) ? 'mixed' : false)
```

File: packages/fluentui/docs/src/examples/components/Checkbox/BestPractices/CheckboxBestPractices.tsx (hunk 7 above)
Comment (assuncaocharles) [reply packages/fluentui/docs/src/examples/components/Checkbox/BestPractices/CheckboxBestPractices.tsx:5]: ?

Comment (layershifter) [reply packages/fluentui/docs/src/examples/components/Checkbox/BestPractices/CheckboxBestPractices.tsx:5]:
```suggestion
const doList = ['Add `aria-controls="id1 id2"` for indeterminate checkbox referring to the controls which it depends'];
```

Do we need `controlsIds` or we can just specify `aria-controls`?

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts [hunk 16]
```diff
@@ -115,9 +117,15 @@ export const checkboxStyles: ComponentSlotStylesPrepared<CheckboxStylesProps, Ch
     backgroundPosition: 'center',
     backgroundRepeat: 'no-repeat',
 
-    ...(p.checked && {
+    ...(p.checked &&
+      p.checked !== 'mixed' && {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts:121]:
```suggestion
      p.checked && {
```

A next condition will win anyway when `checked` is mixed

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts [hunk 17]
```diff
@@ -126,11 +134,22 @@ export const checkboxStyles: ComponentSlotStylesPrepared<CheckboxStylesProps, Ch
     }),
 
     ...(p.disabled &&
-      p.checked && {
+      p.checked &&
+      p.checked !== 'mixed' && {
```
Comment (layershifter) [inline packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts:138]:
```suggestion
       && {
```

Comment (layershifter) [reply packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxStyles.ts:138]: A next condition will win anyway when checked is mixed

File: packages/fluentui/docs/src/examples/components/Checkbox/States/index.tsx [hunk 18]
```diff
@@ -15,6 +15,11 @@ const States = () => (
       description="A checkbox can be read-only and unable to change states."
       examplePath="components/Checkbox/States/CheckboxExampleDisabled"
     />
+    <ComponentExample
+      title="Mixed"
+      description="A checkbox can be mixed."
```
Comment (layershifter) [inline packages/fluentui/docs/src/examples/components/Checkbox/States/index.tsx:20]:
```suggestion
      description="A checkbox can have mixed (indeterminate) state."
```

**Review by layershifter — APPROVED** [review]

**Pull request merged**